File: Python/plot.py
import sys
import matplotlib.pyplot as plt

class axis:
    def __init__(self, lst = []):
        self.data = lst.copy()
        self.size = 0

def plot_y():
    """
    Data piping order:
    1 - size
    2 - data
    3 - xlabel
    4 - ylabel
    5 - title
    """
    y = axis()
    y.size = int(sys.stdin.readline())
    for i in range(y.size):
        y.data.append(float(sys.stdin.readline()))
    
    xlabel = sys.stdin.readline()
    ylabel = sys.stdin.readline()
    title = sys.stdin.readline()

    plt.plot(y.data)
    plt.xlabel(xlabel)
    plt.ylabel(ylabel)
    plt.title(title)
    plt.show()

def plot_xy():
    """
    Data piping order:
    1 - x size
    2 - x data
    3 - y size
    4 - y data
    5 - xlabel
    6 - ylabel
    7 - title
    """
    x = axis()
    y = axis()
    x.size = int(sys.stdin.readline())
    for i in range(x.size):
        x.data.append(float(sys.stdin.readline()))

    y.size = int(sys.stdin.readline())
    for i in range(y.size):
        y.data.append(float(sys.stdin.readline()))
        
    xlabel = sys.stdin.readline()
    ylabel = sys.stdin.readline()
    title = sys.stdin.readline()

    plt.plot(x.data, y.data)
    plt.xlabel(xlabel)
    plt.ylabel(ylabel)
    plt.title(title)
    plt.show()

v = int(sys.stdin.readline())

if (v == 1):
    plot_y()
    print('k')
elif (v == 2):
    plot_xy()

    

File: src/plot.c
/**********************************************
*  plot.c : Plots a signal using pyhon script
*
*  Author: Aduri Sri Sambasiva Advaith
*
***********************************************/
/*
Order to be followed when piping data to be plotted:
1 - size of data
2 - data
3 - xlabel
4 - ylabel
5 - title

Ps:- if multiple data exist, repeat 1,2 before going to 3
*/
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include "../include/signal.h"

char plot_cmd[] = "python3 ../Python/plot.py";

void plot_y(signal sig, char *xlabel, char *ylabel, char *title)
{
    FILE *fp = popen(plot_cmd,"w");
    //sending number of variables to plot
    fprintf(fp,"%d\n",1);
    //sending size of data to plot.py
    fprintf(fp,"%ld\n",sig.size);
    //sending data to plot.py
    for (int i = 0; i < sig.size; i++)
    {
        fprintf(fp,"%f\n",sig.data[i]);
    }
    //sending xlabel to plot.py
    fprintf(fp,"%s\n",xlabel);
    //sending ylabel to plot.py
    fprintf(fp,"%s\n",ylabel);
    //sending title to plot.py
    fprintf(fp,"%s\n",title);
    fclose(fp);
}


void plot_xy(signal x, signal y, char *xlabel, char *ylabel, char *title)
{
    if (x.size != y.size)
    {
        printf("Incorrect signal dimensions.\n");
        return;
    }
    FILE *fp = popen(plot_cmd,"w");
    //sending number of variables to plot
    fprintf(fp,"%d\n",2);
    //sending size of x data to plot.py
    fprintf(fp,"%ld\n",x.size);
    //sending data to plot.py
    for (int i = 0; i < x.size; i++)
    {
        fprintf(fp,"%f\n",x.data[i]);
    }
    //sending size of y data to plot.py
    fprintf(fp,"%ld\n",y.size);
    //sending data to plot.py
    for (int i = 0; i < y.size; i++)
    {
        fprintf(fp,"%f\n",y.data[i]);
    }

    //sending xlabel to plot.py
    fprintf(fp,"%s\n",xlabel);
    //sending ylabel to plot.py
    fprintf(fp,"%s\n",ylabel);
    //sending title to plot.py
    fprintf(fp,"%s\n",title);
    fclose(fp);
}

void plot_ny(signal y[], int size, char *xlabel, char *ylabel, char *title)
{
    FILE *fp = popen(plot_cmd,"w");
    //sending code to plot.py
    fprintf(fp,"%d\n",3);
    //sending number of variables to plot
    fprintf(fp,"%d\n",size);
    for (int i = 0; i < size; i++)
    {
        //sending size of data to plot.py
        fprintf(fp,"%ld\n",y[i].size);
        //sending data to plot.py
        for (int j = 0; j < y[i].size; j++)
        {
            fprintf(fp,"%f\n",y[i].data[j]);
        }
    }
    //sending xlabel to plot.py
    fprintf(fp,"%s\n",xlabel);
    //sending ylabel to plot.py
    fprintf(fp,"%s\n",ylabel);
    //sending title to plot.py
    fprintf(fp,"%s\n",title);
    fclose(fp);
}